design/0001-metallb-crd.md

```diff
@@ -0,0 +1,251 @@
+# MetalLB Custom Resource Controller
+
+<!-- toc -->
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+- [Proposal](#proposal)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Initial CRD definition](#initial-crd-definition)
+- [Testing](#testing)
+- [Debugging and Troubleshooting](#debugging-and-troubleshooting)
+
+<!-- /toc -->
+
+## Summary
+
+The purpose of this enhancement is to be able to use kubernetes Custom
+Resource Definition "CRD" to configure MetalLB layer2 and BGP.
+
+## Motivation
+
+Today MetalLB allows only configuration of layer2 and BGP protocols using
+ConfigMap objects, which is simple from configuration perspective and doesn't
+require cluster-admin rights, however it doesn't have validation on the API
+server level, no support for API versioning and no fine-grained permission
+model via RBAC.
+Custom resource definition "CRD" takes advantage of kubernetes extensibility
+to deliver the automation advantages of cloud services like provisioning,
+scaling and backup/restore.
+
+### Goals
+
+Configuring MetalLB layer2 and BGP using kubernetes custom resource "CRD"
+instead of ConfigMap.
+
+## Proposal
+
+Using SIG's [Kubebuilder](https://github.com/kubernetes-sigs/kubebuilder)
+framework to build new CRD "MetalLB" object and a controller which works as
+a translator from the user's configuration yaml to a MetalLB's configmap
+object, with this approach there will be no changes required for the
+MetalLB management plane.
+
+### Risks and Mitigations
+
+Users need to either use CRD or ConfigMap to provision load balancer instance
+but not both, if user mistakenly used both the CRD configuration will
+overwrite the user defined ConfigMap.
+
+## Design Details
+
+![Alt text](./metallb_crd.png?raw=true "MetalLB Kubernetes Controller")
+
+1. User will supply the yaml configuration using the MetalLB resource.
+1. Configuration will get authenticated and authorized based on RBAC rules.
+1. Supplied configuration will get verified using openAPIV3Schema based on
+   the definition of MetalLB CRD.
+1. New objects will be stored in etcd.
+1. MetalLB controller monitors API server looking for creation of MetalLB
+   resource.
+1. MetalLB controller will use golang based template to convert the configured
+   MetalLB object to ConfigMap object.
+1. The generated ConfigMap object will be written to etcd so MetalLB's
+   components can consume it as if the user directly supplied the ConfigMap
+   configuration.
+
+### Initial CRD definition
+
+In this section will share initial metallB custom resource definition which
+currently covers only layer2, BGP will be added to the same CRD, or we might
+separate it out and use the same controller to manage both resources.
```

Comment:
I think we should either mark this as a work-in-progress and complete the full CRD design before merging, or perhaps split this into 2 PRs.  The first PR could include the high level proposal (use CRDs, use kubebuilder, still support the ConfigMap, ...) but not specify the CRD yet.  The 2nd PR could get into the API details after we reach consensus on persuing CRDs in the first place.

I'm interested in what other people think.  I'm interested in getting consensus on the general approach before working too hard on all of the CRD details.

Reply:
I didn't include the CRD definition originally Mark requested to include, I personally prefer to share it in the next PR with the rest of the code and it will include BGP configuration

Reply:
@russellb I think your suggestion of splitting it makes sense. 1st PR can be a general proposal so we can get consensus on that and the 2nd can be the CRD or a design doc that discusses the CRD.

@msherif1234 My reasoning was that it would be good to agree and merge the CRD as soon as possible and not let development and merging of the full solution hold this up because a) some changes in the CRD may affect the rest of the design and it would avoid any timely rework - although I can't think why that would be the case right now and b) having an agreed and merged CRD may help start other related work earlier if any one is doing any work in parallel.

Reply:
[WIP CRD PR](https://github.com/metallb/metallb/pull/850), the crd yaml is actually auto generated based on the config data structure which is in sync with what Metallb is using I wanted to make metallb use Metallb config spec so we have only one source .